File: other/P28.cpp
//
// Created by enheng on 18-6-10.
//

#include "P28.h"
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    /**
     * @param matrix: matrix, a list of lists of integers
     * @param target: An integer
     * @return: a boolean, indicate whether matrix contains target
     */
    bool searchMatrix(vector<vector<int>> &matrix, int target) {
        // write your code here
        int f = 0, m, b;

        b = matrix.size() - 1;
        if(b == -1) return false;
        if(target < matrix[0][0] || target > matrix[b][matrix[b].size() - 1])
            return false;
        if(target == matrix[b][0]) return true;
        else if(target > matrix[b][0]) return binary_search(matrix[b].begin(), matrix[b].end(), target);
        while(f != b){
            m = (b + f) / 2;
            if(target > matrix[m][0]){
                f = m + 1;
            }else if(target == matrix[m][0]){
                return true;
            }else{
                b = m - 1;
            }
        }

        if(target < matrix[b][0])
            return binary_search(matrix[b - 1].begin(), matrix[b - 1].end(), target);
        else
            return binary_search(matrix[b].begin(), matrix[b].end(), target);
    }
};

void P28(){

}

File: problemID/P28.go
package problemID

import (
    "fmt"
)

/**
 * 搜索二维矩阵
 * @param matrix: matrix, a list of lists of integers
 * @param target: An integer
 * @return: a boolean, indicate whether matrix contains target
 */
func searchMatrix (matrix [][]int, target int) bool {
    // write your code here
    var f, m, b int

    b = len(matrix) - 1
    if b == -1 { return false }
    if target < matrix[0][0] || target > matrix[b][len(matrix[b]) - 1] { return false }
    if target == matrix[b][0] {
        return true
    }else if target > matrix[b][0] {
        return binarySearch2(matrix[b], target)
    }
    for f != b {
        m = (b + f) / 2
        if target > matrix[m][0] {
            f = m + 1
        }else if target == matrix[m][0] {
            return true
        }else{
            b = m - 1
        }
    }

    if target < matrix[b][0] {
        return binarySearch2(matrix[b - 1], target)
    }else {
        return binarySearch2(matrix[f], target)
    }
}

func binarySearch2 (nums []int, target int) bool {
    // write your code here
    var f, m, b int

    b = len(nums) - 1
    for f <= b{
        m = (b + f) / 2
        if target > nums[m] {
            f = m + 1
        }else if target == nums[m] {
            return true
        }else{
            b = m - 1
        }
    }

    return false
}

func P28(){
    fmt.Println(searchMatrix([][]int{}, 15))
}